### ISSUE TITLE
[chore] Test pybind11 2.6.0 (RC 2?)

### ISSUE THREAD

**Issue by henryiii** [pytorch/pytorch#46367]
There [is a bug](https://bugs.python.org/issue41237) in CPython 3.9.0 (already fixed in 3.9 and master) that breaks pybind11 extensions slightly, causing them to exhibit UB (on macOS, usually segfaults on process exit). While this will be fixed in CPython 3.9.1, and has [already been patched](https://github.com/conda-forge/python-feedstock/pull/402) in Conda-forge's Python 3.9 build, it is likely to affect users. pybind11 is about to release 2.6.0 (currently on 2.6.0rc2) that [will detect 3.9.0 and leak instead of segfaulting](https://github.com/pybind/pybind11/pull/2576); can that be tested with PyTorch to make sure it is an easy upgrade and no last fixes or patches are needed? There is at least one long-standing issue requested by PyTorch fixed in 2.6, as well, though I don't remember which one, as well as better CUDA support and more.

**Comment by malfet** [comment]
@henryiii does it happen on any other platforms that OS X?

**Comment by henryiii** [comment]
Only macOS tends to segfault (somewhere between 20-80% of the time), but it's technically UB on all platforms.

**Comment by henryiii** [comment]
Though, in this issue I'm hoping someone who knows how to build pytorch can just bump the pybind11 version to 2.6.0rc2, build, and make sure it doesn't explode or something, at least in a way that we have to fix in pybind11 before release. :)

**Comment by malfet** [comment]
@henryiii do you want to test pybind11 with Python-3.9 or any version of Python will do?

**Comment by henryiii** [comment]
Any version or OS, just to make sure there are no incompatible changes.

**Comment by malfet** [comment]
Will find out soon here: https://github.com/pytorch/pytorch/pull/46415

**Comment by ezyang** [comment]
Last time I tried https://github.com/pytorch/pytorch/pull/31920 ASAN started failing; hope we have better luck this time.

**Comment by henryiii** [comment]
Excellent, glad to see it wasn't the 2.6 changes at least. I was looking through the changes and couldn't see anything in the last several versions that would be likely to fail, but it looks like I wasn't looking back far enough, then.

This _only_ affects llvm, by the way, I rebuild locally in a LLVM docker container and forgot to force LLVM, and it passed. I rebuild forcing LLVM and got the four failures in the CI.

```bash
docker run --rm -v $PWD:/pytorch -it silkeh/clang:9
apt update && apt install -y python3{-dev,-pytest,-pip} ninja-build
cd pytorch/
pip3 install -r requirements.txt
CC=clang CXX=clang++ python3 setup.py develop
(cd test/cpp_extensions/ &&  CC=clang CXX=clang++ python3 setup.py develop)
pytest-3 test/test_cpp_extensions_jit.py
```

I suspect (but am not sure) that the change with CPUDoubleType is due to the extra correctness checking in pybind11 triggering a different pattern of errors than expected, and is easily fixed.

**Comment by malfet** [comment]
@henryiii please let us know if you need any help debugging those

**Comment by henryiii** [comment]
I'd be happy to have some, especially on the first one, I'm fairly familiar with the internals of pybind11 and _using_ PyTorch, but not the internals of PyTorch. :)

Any idea why these would only show up on LLVM?

I think it's okay to move forward on the release of pybind11 2.6.0, and if changes are needed in pybind11, that should be an easy 1.6.1.

**Comment by henryiii** [comment]
What is the JIT mechanism in PyTorch? Is it GCC based, perhaps? In 2.4.0, this was added:
> : Try harder to keep pybind11-internal data structures separate when there are potential ABI incompatibilities.

The compiler ID is now part of the internal structure. I wonder if that is what causes the JIT extension not to find the PyTorch parts?

**Comment by henryiii** [comment]
I suspect it's this commit: https://github.com/pybind/pybind11/commit/c9f5a

**Comment by malfet** [comment]
You are probably right: I'm not 100% sure if pre-CXX11 ABI is still correctly enforce when clang is used as host compiler (because clang on its own is fine, as PR passes CI on OS X)

**Comment by henryiii** [comment]
pybind11 2.6.0rc3 now allows `PYBIND11_COMPILER_TYPE` and `PYBIND11_STDLIB` to be predefined. We could use that to test and see if that really is the problem. What's the best way to do that?

**Comment by malfet** [comment]
Let me update the PR and see if it will fix the issue

**Comment by henryiii** [comment]
You will want to define `PYBIND11_COMPILER_TYPE` and `PYBIND11_STDLIB` to empty strings to restore pre-2.4 behavior. pybind11 gets included in ~20 places or so, so can it be added as a flag in the build system?

**Comment by malfet** [comment]
Yes, this is totally double. Is there a documentation on what those defines are doing?

**Comment by malfet** [comment]
Added defines to the diff, let's see if it will address the problem

**Comment by henryiii** [comment]
Not really, [just here].(https://github.com/pybind/pybind11/blob/c16da993094988141101ac4d96a9b2c92f9ac714/include/pybind11/detail/internals.h#L152-L217)

At the moment, it probably should be considered a secret backdoor that lets you override the ABI string.

**Comment by henryiii** [comment]
It needs to be an empty string, currently it's `1`. It's going to get concatenated into a ABI strong.
